### PR TITLE
Issue #4004 AssertionError from Range.gap.

### LINKED ISSUE #4004: AssertionError from Range.gap
@zhanhb [reports](https://github.com/google/guava/commit/a9dd7098f741d652fe2daf8193b831737dc36aa0#commitcomment-41940819):

```java
Range.atLeast(1).gap(Range.atLeast(2));
```
```txt
java.lang.AssertionError
	at com.google.common.collect.Cut$AboveAll.describeAsLowerBound(Cut.java:259)
	at com.google.common.collect.Range.toString(Range.java:674)
	at com.google.common.collect.Range.<init>(Range.java:357)
	at com.google.common.collect.Range.create(Range.java:155)
	at com.google.common.collect.Range.gap(Range.java:582)
```

### PR DESCRIPTION
**Pull request by Koooooo-7**
Fix #4004, Got IAE instead of AssertionError.

### CODE REVIEW

**Comment by Koooooo-7** [comment]
@googlebot I signed it!

**Comment by cpovirk** [comment]
Thanks. I suspect that what we'd prefer is to avoid calling `describeAsLowerBound` at all in this case. But I'd have to look at `gap` more to confirm (and to figure out what to do instead).

**Comment by Koooooo-7** [comment]
When the invalid range is `(+∞, 2)`，it will throw the IAE, it make sense. but the  ` toString(lowerBound, upperBound)` throws exception again.
```java
  private Range(Cut<C> lowerBound, Cut<C> upperBound) {
    this.lowerBound = checkNotNull(lowerBound);
    this.upperBound = checkNotNull(upperBound);
    if (lowerBound.compareTo(upperBound) > 0
        || lowerBound == Cut.<C>aboveAll()
        || upperBound == Cut.<C>belowAll()) {
      throw new IllegalArgumentException("Invalid range: " + toString(lowerBound, upperBound));
    }
  }
```
Cuz the `+∞`  called the `Cut$AboveAll#describeAsLowerBound()`, throw the AssertionError.
```java
   @Override
    void describeAsLowerBound(StringBuilder sb) {
      throw new AssertionError();
    }
```
IMHO, another way to solve it is to make `describeAsLowerBound()` return `+∞` like what the  `Cut$belowAll` did now.
```java
 @Override
    void describeAsLowerBound(StringBuilder sb) {
      sb.append("(-\u221e");
    }
```
```java
 @Override
    void describeAsLowerBound(StringBuilder sb) {
      sb.append("(+\u221e");;
    }
```
 then, the exception should like this .
```shell
java.lang.IllegalArgumentException: Invalid range: (+∞..2)

	at com.google.common.collect.Range.<init>(Range.java:358)
	at com.google.common.collect.Range.create(Range.java:156)
```
WDYT?

**Comment by Koooooo-7** [comment]
Besides that , I made an opposite case.
```java 
Range.atMost(3).gap(Range.atMost(2));
```
```shell
java.lang.AssertionError
	at com.google.common.collect.Cut$BelowAll.describeAsUpperBound(Cut.java:165)
	at com.google.common.collect.Range.toString(Range.java:677)
	at com.google.common.collect.Range.<init>(Range.java:358)
	at com.google.common.collect.Range.create(Range.java:156)
	at com.google.common.collect.Range.gap(Range.java:583)
```
it seems the similar issue at `Cut$BelowAll.describeAsUpperBound`  also.

**Comment by Koooooo-7** [comment]
According to the rules, I added the pre range checks for  `gap`.
```java
   * <p>The gap exists if and only if the two ranges are either disconnected or immediately adjacent
   * (any intersection must be an empty range).
```
PTAL, @cpovirk

**Comment by cpovirk** [comment]
Sounds reasonable, thanks. Would you also:

- Add your tests for the AssertionError cases

- include the 2 input ranges in the exception message

**Comment by Koooooo-7** [comment]
Yup, I will add the tests. and I found that the problem is only when the gap result is an invalid range which contains `+∞`/`-∞`. f.e.
```java
Range.atLeast(1).gap(Range.atLeast(2));
```
the invalid gap range result is `(+∞, 2)`,  it will try to get the IAE .
```java
 throw new IllegalArgumentException("Invalid range: " + toString(lowerBound, upperBound));
```
but in the IAE, it will call the `toString(lowerBound, upperBound)` and get the the `AssertionError` when you wanna get the input ranges in the exception message.
```java
 private static String toString(Cut<?> lowerBound, Cut<?> upperBound) {
    StringBuilder sb = new StringBuilder(16);
    lowerBound.describeAsLowerBound(sb);
    sb.append("..");
    upperBound.describeAsUpperBound(sb);
    return sb.toString();
  }
```
The reason it the  `+∞`  will call the `Cut$AboveAll#describeAsLowerBound()`, throw the AssertionError.
```java
   @Override
    void describeAsLowerBound(StringBuilder sb) {
      throw new AssertionError();
    }
```
IMHO, another way to solve it is to make `describeAsLowerBound()` return `+∞` . cuz when you wanna call  `toString(lowerBound, upperBound)` to get the invalid ranges, the `(+∞, 2)` or `(1, -∞)` r both make sense.
```java
   @Override
    void describeAsLowerBound(StringBuilder sb) {
      sb.append("(+\u221e");
    }
```
WDYT?

**Comment by Koooooo-7** [comment]
@cpovirk  I added the test and the invalid ranges in the IAE message.

**Comment by cpovirk** [comment]
Thanks. I got another set of eyes on this, and hopefully we'll get it submitted internally and merged out soon.

File: guava-tests/test/com/google/common/collect/RangeTest.java [hunk 1]
```diff
@@ -472,6 +472,20 @@ public void testGap_overlapping() {
...
 
+  public void testGap_invalidRangesWithInfinity(){
+
+    try {
+      Range.atLeast(1).gap(Range.atLeast(2));
+    } catch (IllegalArgumentException expected) {
+    }
+
+    try {
+      Range.atMost(1).gap(Range.atMost(2));
```
[the remark this answers is not in the document]
Comment (Koooooo-7) [reply guava-tests/test/com/google/common/collect/RangeTest.java:483]: It does throw an IAE exception and the tests passed well locally.

**Pull request closed without merging**